<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Workshop on bacterial comparative genomics"><meta name=author content="Jacques Dainat et al."><link href=https://southgreenplatform.github.io/training_bacterial_comparative_genomics/pages/training/ rel=canonical><link rel=prev href=../..><link href=../lecture/ rel=next><link rel=icon href=../../assets/images/favicon.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.6.14"><title>Tutorials - Training on bacterial comparative genomics</title><link rel=stylesheet href=../../assets/stylesheets/main.342714a4.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#handsome-bacterial-comparative-genomics class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="Training on bacterial comparative genomics" class="md-header__button md-logo" aria-label="Training on bacterial comparative genomics" data-md-component=logo> <img src=../images/southgreen.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Training on bacterial comparative genomics </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Tutorials </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/SouthGreenPlatform/training_bacterial_comparative_genomics title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> training_bacterial_comparative_genomics </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Home </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=./ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../lecture/ class=md-tabs__link> Lecture </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="Training on bacterial comparative genomics" class="md-nav__button md-logo" aria-label="Training on bacterial comparative genomics" data-md-component=logo> <img src=../images/southgreen.png alt=logo> </a> Training on bacterial comparative genomics </label> <div class=md-nav__source> <a href=https://github.com/SouthGreenPlatform/training_bacterial_comparative_genomics title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> training_bacterial_comparative_genomics </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Tutorials </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#first-commands-for-preparing-the-working-environment class=md-nav__link> <span class=md-ellipsis> First commands for preparing the working environment </span> </a> </li> <li class=md-nav__item> <a href=#genome-assembly-from-oxford-nanopore-technologies-ont-long-reads-using-flye class=md-nav__link> <span class=md-ellipsis> Genome Assembly (from Oxford Nanopore Technologies (ONT) long reads) (using Flye) </span> </a> </li> <li class=md-nav__item> <a href=#separate-chromosomal-and-plasmid-scaffolds-using-mob-suite class=md-nav__link> <span class=md-ellipsis> Separate chromosomal and plasmid scaffolds (using MOB-Suite) </span> </a> </li> <li class=md-nav__item> <a href=#genome-annotation-using-prokka class=md-nav__link> <span class=md-ellipsis> Genome annotation (using Prokka) </span> </a> </li> <li class=md-nav__item> <a href=#gc-content-analysis-skewit class=md-nav__link> <span class=md-ellipsis> GC content analysis (SkewIT) </span> </a> </li> <li class=md-nav__item> <a href=#visualize-genome-annotation-using-circos class=md-nav__link> <span class=md-ellipsis> Visualize genome annotation (using Circos) </span> </a> </li> <li class=md-nav__item> <a href=#retrieve-public-genomes-available-for-comparison class=md-nav__link> <span class=md-ellipsis> Retrieve public genomes available for comparison </span> </a> </li> <li class=md-nav__item> <a href=#convert-genbank-annotation-files-to-gff-format class=md-nav__link> <span class=md-ellipsis> Convert GenBank annotation files to GFF format </span> </a> </li> <li class=md-nav__item> <a href=#comparison-between-2-genomes-with-minimap2-and-circos class=md-nav__link> <span class=md-ellipsis> Comparison between 2 genomes (with Minimap2 and circos) </span> </a> </li> <li class=md-nav__item> <a href=#compare-genome-similarity-using-average-nucleotide-identity-ani-fastani class=md-nav__link> <span class=md-ellipsis> Compare genome similarity using Average Nucleotide Identity (ANI) (fastANI) </span> </a> </li> <li class=md-nav__item> <a href=#gene-content-comparison-pangenome-analysis-using-roary class=md-nav__link> <span class=md-ellipsis> Gene content comparison / Pangenome analysis (using Roary) </span> </a> </li> <li class=md-nav__item> <a href=#association-tests-with-metadata-pan-gwas-with-scoary class=md-nav__link> <span class=md-ellipsis> Association tests with metadata (Pan-GWAS) (with Scoary) </span> </a> </li> <li class=md-nav__item> <a href=#explore-pan-genome-with-panexplorer-web-application class=md-nav__link> <span class=md-ellipsis> Explore pan-genome with PanExplorer web application </span> </a> </li> <li class=md-nav__item> <a href=#construct-a-pangenome-graph-using-minigraph2 class=md-nav__link> <span class=md-ellipsis> Construct a pangenome graph (using minigraph2) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../lecture/ class=md-nav__link> <span class=md-ellipsis> Lecture </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/SouthGreenPlatform/training_bacterial_comparative_genomics/edit/main/docs/pages/training.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg> </a> <a href=https://github.com/SouthGreenPlatform/training_bacterial_comparative_genomics/raw/main/docs/pages/training.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg> </a> <h1 id=handsome-bacterial-comparative-genomics>Handsome bacterial comparative genomics<a class=headerlink href=#handsome-bacterial-comparative-genomics title="Permanent link">#</a></h1> <h2 id=first-commands-for-preparing-the-working-environment><strong>First commands for preparing the working environment</strong><a class=headerlink href=#first-commands-for-preparing-the-working-environment title="Permanent link">#</a></h2> <p>First create your working directory:</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg
<span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg
</code></pre></div> <p>We can now download the git repository of this training that contains material needed for the training.</p> <div class=highlight><pre><span></span><code>git<span class=w> </span>clone<span class=w> </span>https://github.com/SouthGreenPlatform/training_bacterial_comparative_genomics.git
</code></pre></div> <p>We also need some raw data:</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>raw_data
<span class=nb>cd</span><span class=w> </span>raw_data
wget<span class=w> </span>https://panexplorer.southgreen.fr/data/mysample.fastq.gz
</code></pre></div> <div class="admonition question"> <p class=admonition-title>Question</p> <pre><code>Using a unix command, count how many reads are in the raw Fastq file?
</code></pre> </div> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>zgrep<span class=w> </span>-c<span class=w> </span><span class=s1>&#39;barcode=&#39;</span><span class=w> </span>mysample.fastq.gz
</code></pre></div> </details> <h2 id=genome-assembly-from-oxford-nanopore-technologies-ont-long-reads-using-flye><strong>Genome Assembly (from Oxford Nanopore Technologies (ONT) long reads) (using Flye)</strong><a class=headerlink href=#genome-assembly-from-oxford-nanopore-technologies-ont-long-reads-using-flye title="Permanent link">#</a></h2> <p>We start by creating and moving into a directory dedicated for the task:</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/assembly
<span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/assembly
</code></pre></div> <p>We will use Flye to perform the genome assembly. <br> Load the appropriate module for flye (module load) or install the tool locally with conda </p> <div class=highlight><pre><span></span><code>conda<span class=w> </span>create<span class=w> </span>-n<span class=w> </span>flye<span class=w> </span>-c<span class=w> </span>bioconda<span class=w> </span>flye
conda<span class=w> </span>activate<span class=w> </span>flye
</code></pre></div> <p>We can now run the assembly </p> <div class=highlight><pre><span></span><code>flye<span class=w> </span>--nano-raw<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/raw_data/mysample.fastq.gz<span class=w> </span>-o<span class=w> </span>assembly<span class=w> </span>&gt;&gt;flye.log<span class=w> </span><span class=m>2</span>&gt;<span class=p>&amp;</span><span class=m>1</span>
</code></pre></div> <p>As this task is a high time consuming step, we can stop the tool with <code>CTRL-C</code> and download the expected output:</p> <div class=highlight><pre><span></span><code>cp<span class=w> </span>-rf<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/training_bacterial_comparative_genomics/data/precomputed_assembly/assembly_precomputed.fasta<span class=w> </span>.
</code></pre></div> <div class="admonition question"> <p class=admonition-title>Question</p> <pre><code>How many contigs could be assembled from reads?
</code></pre> </div> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>grep<span class=w> </span>-c<span class=w> </span><span class=s2>&quot;&gt;&quot;</span><span class=w> </span>assembly_precomputed.fasta
</code></pre></div> </details> <h2 id=separate-chromosomal-and-plasmid-scaffolds-using-mob-suite><strong>Separate chromosomal and plasmid scaffolds (using MOB-Suite)</strong><a class=headerlink href=#separate-chromosomal-and-plasmid-scaffolds-using-mob-suite title="Permanent link">#</a></h2> <p>Let's start by creating and moving into a directory dedicated for the task:</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/mob_recon
<span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/mob_recon
</code></pre></div> <p>We will run MOB-suite using a singularity image. </p> <p>Run the <code>mob_recon</code> utility using singularity image (<code>/scratch3/TP_bcg/mob_suite_3.0.3.sif</code>) to separate chromosome and plasmid contigs<br> First, load the appropriate module for running singularity Use the <code>--bind</code> option to bind your current directory on the host to <code>/mnt</code> in the container</p> <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>system/singularity/3.6.0
cp<span class=w> </span>-rf<span class=w> </span>/scratch3/TP_bcg/mob_suite_3.0.3.sif<span class=w> </span>.
cp<span class=w> </span>-rf<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/training_bacterial_comparative_genomics/data/precomputed_assembly/assembly_precomputed.fasta<span class=w> </span>.
singularity<span class=w> </span><span class=nb>exec</span><span class=w> </span>--bind<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/mob_recon/:/tmp<span class=w> </span>mob_suite_3.0.3.sif<span class=w> </span>mob_recon<span class=w> </span>-i<span class=w> </span>/tmp/assembly_precomputed.fasta<span class=w> </span>-o<span class=w> </span>/tmp/assembly.mob_recon
</code></pre></div> <p>The results is made by 4 files <code>chromosome.fasta contig_report.txt mobtyper_results.txt plasmid_AD399.fasta</code></p> <p>We can have a look at the report:</p> <div class="admonition question"> <p class=admonition-title>Question</p> <pre><code>Looking at the MOB-SUITE report, does our sequenced sample contain plasmid sequence?
</code></pre> </div> <details class=example> <summary>Solution</summary> <p><div class=highlight><pre><span></span><code>more<span class=w> </span>assembly.mob_recon/contig_report.txt
</code></pre></div> <div class=highlight><pre><span></span><code>sample_id   molecule_type   primary_cluster_id  secondary_cluster_id    contig_id   size    gc  md5 circularity_status  rep_type(s) rep_type_accession(s)   relaxase_type(s)    relaxase_type_accession(s)  mpf_type    mpf_type_accession(s)   orit_type(s)    orit_accession(s)   predicted_mobility  mash_nearest_neighbor   mash_neighbor_distance  mash_neighbor_identification    repetitive_dna_id   repetitive_dna_type filtering_reason
assembly_precomputed    chromosome  -   -   contig_1_circular_rotated   4682534 63.833983907004196  cf955654136e9041810a80b8191fa41f    not tested  -   -   MOBP    NC_007507_00032 -   -   -   -   -   -   -   -   -   -   -
assembly_precomputed    plasmid AD399   -   contig_2_circular_rotated   69058   61.861044339540676  475037e138b4ade1836aa05f421c47b8    not tested  rep_cluster_1289    000607__CP000620_00033  MOBP    CP022994_00148  -   -   -   -   -   CP033195    0.0537328   Xanthomonas oryzae pv. oryzae   -   -   -
</code></pre></div></p> </details> <h2 id=genome-annotation-using-prokka><strong>Genome annotation (using Prokka)</strong><a class=headerlink href=#genome-annotation-using-prokka title="Permanent link">#</a></h2> <p>We will now annotate the chromosome. Let's start by creating and moving into a directory dedicated for the task. </p> <p>Annotation will be performed on the chromosome file only (not plasmid). A symbolic link can be created to have the chromosome.fasta file in the current directory.</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/prokka
<span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/prokka
ln<span class=w> </span>-s<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/mob_recon/assembly.mob_recon/chromosome.fasta
</code></pre></div> <p>We will use Prokka to perform the genome annotation. Load the appropriate module for prokka (module load) or install the tool locally with conda</p> <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>bioinfo/prokka/1.14.6
</code></pre></div> <p>or</p> <div class=highlight><pre><span></span><code>conda<span class=w> </span>create<span class=w> </span>-n<span class=w> </span>prokka<span class=w> </span>-c<span class=w> </span>conda-forge<span class=w> </span>-c<span class=w> </span>bioconda<span class=w> </span>prokka
conda<span class=w> </span>activate<span class=w> </span>prokka
</code></pre></div> <p>List the database that will be used by prokka for annotation</p> <div class=highlight><pre><span></span><code>prokka --listdb
</code></pre></div> <p>We can now launch the annotation: </p> <div class=highlight><pre><span></span><code>prokka<span class=w> </span>chromosome.fasta<span class=w> </span>--prefix<span class=w> </span>assembly<span class=w> </span>--force<span class=w> </span>--outdir<span class=w> </span>prokka_out<span class=w> </span>&gt;&gt;<span class=w> </span>prokka.log<span class=w> </span><span class=m>2</span>&gt;<span class=p>&amp;</span><span class=m>1</span>
</code></pre></div> <p>You can now deactivate the prokka conda environment</p> <div class=highlight><pre><span></span><code>conda<span class=w> </span>deactivate
</code></pre></div> <p>Let's have a look at the Genebank output file:</p> <div class=highlight><pre><span></span><code>head<span class=w> </span>-40<span class=w> </span>prokka_out/assembly.gbk
</code></pre></div> <details class=example> <summary>Solution</summary> <pre><code>LOCUS       contig_1_circular_rotated4682534 bp   DNA  linear       10-JUN-2022
DEFINITION  Genus species strain strain.
ACCESSION   
VERSION
KEYWORDS    .
SOURCE      Genus species
ORGANISM  Genus species
            Unclassified.
COMMENT     Annotated using prokka 1.14.6 from
            https://github.com/tseemann/prokka.
FEATURES             Location/Qualifiers
    source          1..4682534
                    /organism="Genus species"
                    /mol_type="genomic DNA"
                    /strain="strain"
    CDS             1..1329
                    /gene="dnaA"
                    /locus_tag="GBGACGDJ_00001"
                    /inference="ab initio prediction:Prodigal:002006"
                    /inference="similar to AA sequence:UniProtKB:P03004"
                    /codon_start=1
                    /transl_table=11
                    /product="Chromosomal replication initiator protein DnaA"
                    /db_xref="COG:COG0593"
                    /translation="MDAWPRCLERLEAEFPPEDVHTWLKPLQAEDRGDSIVLYAPNAF
                    IVDQVRERYLPRIRELLAYFAGNREVALAVGSRPRAPEPEPAPVAATIAPQAAPIAPF
                    AGNLDSHYTFANFVEGRSNQLGLAAAIQAAQKPGDRAHNPLLLYGSTGLGKTHLMFAA
                    GNALRQAKPAAKVMYLRSEQFFSAMIRALQDKAMDQFKRQFQQIDALLIDDIQFFAGK
                    DRTQEEFFHTFNALFDGRQQIILTCDRYPREVEGLEPRLKSRLAWGLSVAIDPPDFET
                    RAAIVLAKARERGAEIPDDVAFLIAKKMRSNVRDLEGALNTLVARANFTGRSITVEFA
                    QETLRDLLRAQQQAIGIPNIQKTVADYYGLQMKDLLSKRRTRSLARPRQVAMALAKEL
                    TEHSLPEIGDAFAGRDHTTVLHACRQIRTLMEADGKLREDWEKLIRKLSE"
    CDS             1607..2389
                    /gene="dnaN_1"
                    /locus_tag="GBGACGDJ_00002"
                    /inference="ab initio prediction:Prodigal:002006"
                    /inference="similar to AA sequence:UniProtKB:Q9I7C4"
                    /codon_start=1
                    /transl_table=11
                    /product="Beta sliding clamp"
</code></pre> </details> <p>Now let's have a look at the GFF output file:</p> <div class=highlight><pre><span></span><code>head<span class=w> </span>-10<span class=w> </span>prokka_out/assembly.gff
</code></pre></div> <details class=example> <summary>Solution</summary> <pre><code>##gff-version 3
##sequence-region contig_1_circular_rotated 1 4682534
contig_1_circular_rotated   Prodigal:002006 CDS 1   1329    .   +   0   ID=GBGACGDJ_00001;Name=dnaA;db_xref=COG:COG0593;gene=dnaA;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P03004;locus_tag=GBGACGDJ_00001;product=Chromosomal replication initiator protein DnaA
contig_1_circular_rotated   Prodigal:002006 CDS 1607    2389    .   +   0   ID=GBGACGDJ_00002;Name=dnaN_1;db_xref=COG:COG0592;gene=dnaN_1;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:Q9I7C4;locus_tag=GBGACGDJ_00002;product=Beta sliding clamp
contig_1_circular_rotated   Prodigal:002006 CDS 2346    2708    .   +   0   ID=GBGACGDJ_00003;Name=dnaN_2;db_xref=COG:COG0592;gene=dnaN_2;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P0A988;locus_tag=GBGACGDJ_00003;product=Beta sliding clamp
contig_1_circular_rotated   Prodigal:002006 CDS 3777    4883    .   +   0   ID=GBGACGDJ_00004;Name=recF;db_xref=COG:COG1195;gene=recF;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P0A7H0;locus_tag=GBGACGDJ_00004;product=DNA replication and repair protein RecF
contig_1_circular_rotated   Prodigal:002006 CDS 4999    7443    .   +   0   ID=GBGACGDJ_00005;eC_number=5.6.2.2;Name=gyrB;db_xref=COG:COG0187;gene=gyrB;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P0A2I3;locus_tag=GBGACGDJ_00005;product=DNA gyrase subunit B
contig_1_circular_rotated   Prodigal:002006 CDS 7511    8347    .   +   0   ID=GBGACGDJ_00006;inference=ab initio prediction:Prodigal:002006;locus_tag=GBGACGDJ_00006;product=hypothetical protein
contig_1_circular_rotated   Prodigal:002006 CDS 8579    9340    .   +   0   ID=GBGACGDJ_00007;eC_number=3.4.-.-;Name=bepA_1;gene=bepA_1;inference=ab initio prediction:Prodigal:002006,protein motif:HAMAP:MF_00997;locus_tag=GBGACGDJ_00007;product=Beta-barrel assembly-enhancing protease
contig_1_circular_rotated   Prodigal:002006 CDS 9617    10810   .   +   0   ID=GBGACGDJ_00008;inference=ab initio prediction:Prodigal:002006;locus_tag=GBGACGDJ_00008;product=hypothetical protein
</code></pre> </details> <p>Check that prokka assign a functional COG annotation in genbank or gff files.</p> <div class=highlight><pre><span></span><code>grep<span class=w> </span>COG<span class=w> </span>prokka_out/assembly.gff<span class=w> </span><span class=p>|</span><span class=w> </span>head<span class=w> </span>-5
</code></pre></div> <details class=example> <summary>Solution</summary> <pre><code>contig_1_circular_rotated   Prodigal:002006 CDS 1   1329    .   +   0   ID=GBGACGDJ_00001;Name=dnaA;db_xref=COG:COG0593;gene=dnaA;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P03004;locus_tag=GBGACGDJ_00001;product=Chromosomal replication initiator protein DnaA
contig_1_circular_rotated   Prodigal:002006 CDS 1607    2389    .   +   0   ID=GBGACGDJ_00002;Name=dnaN_1;db_xref=COG:COG0592;gene=dnaN_1;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:Q9I7C4;locus_tag=GBGACGDJ_00002;product=Beta sliding clamp
contig_1_circular_rotated   Prodigal:002006 CDS 2346    2708    .   +   0   ID=GBGACGDJ_00003;Name=dnaN_2;db_xref=COG:COG0592;gene=dnaN_2;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P0A988;locus_tag=GBGACGDJ_00003;product=Beta sliding clamp
contig_1_circular_rotated   Prodigal:002006 CDS 3777    4883    .   +   0   ID=GBGACGDJ_00004;Name=recF;db_xref=COG:COG1195;gene=recF;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P0A7H0;locus_tag=GBGACGDJ_00004;product=DNA replication and repair protein RecF
contig_1_circular_rotated   Prodigal:002006 CDS 4999    7443    .   +   0   ID=GBGACGDJ_00005;eC_number=5.6.2.2;Name=gyrB;db_xref=COG:COG0187;gene=gyrB;inference=ab initio prediction:Prodigal:002006,similar to AA sequence:UniProtKB:P0A2I3;locus_tag=GBGACGDJ_00005;product=DNA gyrase subunit B
</code></pre> </details> <p>Using awk for extracting information, prepare two files named "genes_plus.txt" and "genes_minus.txt" for representing genes with Circos, located respectively in positive and negative strand. Circos input file must respect the following format with space separator<br> <code>Chr start end</code></p> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>awk<span class=w> </span><span class=o>{</span><span class=s1>&#39;if ($3 == &quot;CDS&quot; &amp;&amp; $7 == &quot;+&quot;)print &quot;Chr &quot;$4&quot; &quot;$5&#39;</span><span class=o>}</span><span class=w> </span>prokka_out/assembly.gff<span class=w> </span>&gt;genes_plus.txt
awk<span class=w> </span><span class=o>{</span><span class=s1>&#39;if ($3 == &quot;CDS&quot; &amp;&amp; $7 == &quot;-&quot;)print &quot;Chr &quot;$4&quot; &quot;$5&#39;</span><span class=o>}</span><span class=w> </span>prokka_out/assembly.gff<span class=w> </span>&gt;genes_minus.txt
</code></pre></div> </details> <h2 id=gc-content-analysis-skewit><strong>GC content analysis (SkewIT)</strong><a class=headerlink href=#gc-content-analysis-skewit title="Permanent link">#</a></h2> <p>We will now use SkewIT for analyzing GC Skew. </p> <p>GC skew is a statistical method for measuring strand-specific guanine overrepresentation.<br> The GC skew is proven to be useful as the indicator of the DNA leading strand, lagging strand, replication origin, and replication terminal.<br> The GC skew is positive and negative in the leading strand and in the lagging strand respectively; therefore, it is expected to see a switch in GC skew sign just at the point of DNA replication origin and terminus</p> <p>GC skew = (G - C)/(G + C)</p> <p>Let's start by creating and moving into a directory dedicated for the task:</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/skewit
<span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/skewit
ln<span class=w> </span>-s<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/mob_recon/chromosome.fasta
</code></pre></div> <p>Let's start by installing the tool:</p> <div class=highlight><pre><span></span><code>git<span class=w> </span>clone<span class=w> </span>https://github.com/jenniferlu717/SkewIT.git
</code></pre></div> <p>We can now run the analysis:</p> <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>system/python/3.8.12
python<span class=w> </span>SkewIT/src/gcskew.py<span class=w> </span>-i<span class=w> </span>chromosome.fasta<span class=w> </span>-o<span class=w> </span>gcskew.txt<span class=w> </span>-k<span class=w> </span><span class=m>500</span>
</code></pre></div> <p>Let's have a look at the result:</p> <div class=highlight><pre><span></span><code>head<span class=w> </span>-5<span class=w> </span>gcskew.txt
</code></pre></div> <details class=example> <summary>Solution</summary> <pre><code>Sequence    Index   GC Skew (0kb)
contig_1_circular_rotated   0   -0.05952381
contig_1_circular_rotated   500 0.00958466
contig_1_circular_rotated   1000    -0.00327869
contig_1_circular_rotated   1500    -0.01010101
</code></pre> </details> <p>Using awk, prepare a file named "gcskew.circos.txt" for representing GC skew with Circos <br> Circos input file must respect the following format with space separator<br> <code>Chr position value</code> </p> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>grep<span class=w> </span>-v<span class=w> </span><span class=s1>&#39;Sequence&#39;</span><span class=w> </span>gcskew.txt<span class=w> </span><span class=p>|</span><span class=w> </span>awk<span class=w> </span><span class=o>{</span><span class=s1>&#39;print &quot;Chr &quot;$2&quot; &quot;$2+500&quot; &quot;$3&#39;</span><span class=o>}</span><span class=w> </span>&gt;<span class=w> </span>gcskew.circos.txt
</code></pre></div> </details> <h2 id=visualize-genome-annotation-using-circos><strong>Visualize genome annotation (using Circos)</strong><a class=headerlink href=#visualize-genome-annotation-using-circos title="Permanent link">#</a></h2> <p>Create a directory for circos and move into this directory dedicated for the task.</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/circos
<span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/circos
</code></pre></div> <p>Install Circos by cloning the GitHub repository</p> <div class=highlight><pre><span></span><code>git<span class=w> </span>clone<span class=w> </span>https://github.com/vigsterkr/circos.git
<span class=nb>cd</span><span class=w> </span>circos
./install-unix
<span class=nb>cd</span><span class=w> </span>..
</code></pre></div> <p>Create a karyotype file for Circos, indicating the name, size (4.6Mb) and color of the chromosome.</p> <div class=highlight><pre><span></span><code><span class=nb>echo</span><span class=w> </span><span class=s2>&quot;chr - Chr 1 0 4600000 black&quot;</span><span class=w> </span>&gt;karyotype.txt
</code></pre></div> <p>Copy the template of circos configuration file located in the training material.</p> <div class=highlight><pre><span></span><code>cp -rf /scratch3/TP_bcg/$USER/training_bcg/training_bacterial_comparative_genomics/data/circos1.conf ./circos1.conf
</code></pre></div> <p>Edit the Circos configuration file (circos1.conf) to adapt and customize your Circos image. </p> <p>Launch circos by specifying your circos configuration file with <code>-conf</code> option</p> <div class=highlight><pre><span></span><code>circos/bin/circos<span class=w> </span>-conf<span class=w> </span>circos1.conf
</code></pre></div> <p><div class=highlight><pre><span></span><code>cp<span class=w> </span>-rf<span class=w> </span>circos.png<span class=w> </span>circos1.png
</code></pre></div> This should result in a circular representation of genes along your chromosome as two distinct tracks.<br> An additional track shows the GC skew.</p> <p><img alt src=../images/circos1.svg></p> <div class="admonition question"> <p class=admonition-title>Exercise</p> <pre><code>Follow the same process to include a new track for tRNA in your Circos
</code></pre> </div> <h2 id=retrieve-public-genomes-available-for-comparison><strong>Retrieve public genomes available for comparison</strong><a class=headerlink href=#retrieve-public-genomes-available-for-comparison title="Permanent link">#</a></h2> <p>Public genomes can be retrived using NCBI : <a href=https://www.ncbi.nlm.nih.gov/datasets/genomes/ >https://www.ncbi.nlm.nih.gov/datasets/genomes/</a></p> <div class="admonition question"> <p class=admonition-title>Questions</p> <pre><code>How many public genomes of Xanthomonas have been released in Genbank?

How many genomes are complete?
</code></pre> </div> <p>Alternatively, one can use the remote file to explore Genbank releases of genomes. <a href=https://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/prokaryotes.txt>https://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/prokaryotes.txt</a></p> <p>Let's start by creating a new directory "public_genomes" and moving into this directory dedicated for the task:</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/public_genomes
<span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/public_genomes
</code></pre></div> <p>Using the wget command, retrieve the prokaryotes.txt file hosted at NCBI.</p> <div class="admonition question"> <p class=admonition-title>Questions</p> <pre><code>Can we find the same number of Xanthomonas genomes from this file?  
How many public genomes of Xanthomonas oryzae species have been released in Genbank?  
How many genomes are complete?  
How many public complete genomes of Xanthomonas oryzae, pathovar oryzae are available?
</code></pre> </div> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/prokaryotes.txt
</code></pre></div> <div class=highlight><pre><span></span><code>grep<span class=w> </span><span class=s2>&quot;Xanthomonas&quot;</span><span class=w> </span>prokaryotes.txt
</code></pre></div> <div class=highlight><pre><span></span><code>grep<span class=w> </span><span class=s2>&quot;Xanthomonas&quot;</span><span class=w> </span>prokaryotes.txt<span class=w> </span><span class=p>|</span><span class=w> </span>grep<span class=w> </span>-c<span class=w> </span><span class=s1>&#39;chromosome:&#39;</span>
</code></pre></div> <div class=highlight><pre><span></span><code>grep<span class=w> </span><span class=s2>&quot;Xanthomonas oryzae pv. oryzae&quot;</span><span class=w> </span>prokaryotes.txt<span class=w> </span><span class=p>|</span><span class=w> </span>grep<span class=w> </span>-c<span class=w> </span><span class=s1>&#39;chromosome:&#39;</span>
</code></pre></div> <div class=highlight><pre><span></span><code>grep<span class=w> </span><span class=s2>&quot;Xanthomonas oryzae pv. oryzicola&quot;</span><span class=w> </span>prokaryotes.txt<span class=w> </span><span class=p>|</span><span class=w> </span>grep<span class=w> </span>-c<span class=w> </span><span class=s1>&#39;chromosome:&#39;</span>
</code></pre></div> </details> <p>Using metadata contained in the file, retrieve the genome files (fasta and annotations) of the first 3 complete genome of Xoo (Xanthomonas oryzae pv. oryzae) (as they appear in the file).<br> Name each genome file with the prefix Xoo_</p> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>grep<span class=w> </span><span class=s2>&quot;Xanthomonas oryzae pv. oryzicola&quot;</span><span class=w> </span>prokaryotes.txt<span class=w> </span><span class=p>|</span><span class=w> </span>grep<span class=w> </span><span class=s1>&#39;chromosome:&#39;</span><span class=w> </span><span class=p>|</span><span class=w> </span>head<span class=w> </span>-3<span class=w> </span>
Xanthomonas<span class=w> </span>oryzae<span class=w> </span>pv.<span class=w> </span>oryzicola<span class=w>    </span><span class=m>129394</span><span class=w>  </span>PRJNA248159<span class=w> </span><span class=m>248159</span><span class=w>  </span>Proteobacteria<span class=w>  </span>Gammaproteobacteria<span class=w> </span><span class=m>4</span>.43083<span class=w> </span><span class=m>64</span>.1<span class=w>    </span>chromosome:NZ_CP007810.1/CP007810.1<span class=w> </span>-<span class=w>   </span><span class=m>1</span><span class=w>   </span><span class=m>4137</span><span class=w>    </span><span class=m>3486</span><span class=w>    </span><span class=m>2015</span>/06/08<span class=w>  </span><span class=m>2021</span>/11/23<span class=w>  </span>Complete<span class=w> </span>Genome<span class=w> </span>Key<span class=w> </span>Laboratory<span class=w> </span>of<span class=w> </span>Agricultural<span class=w> </span>Biodiversity<span class=w> </span><span class=k>for</span><span class=w> </span>Plant<span class=w> </span>Disease<span class=w> </span>Management<span class=w> </span>under<span class=w> </span>the<span class=w> </span>Ministry<span class=w> </span>of<span class=w> </span>Education<span class=w>    </span>SAMN02793996<span class=w>    </span>GCA_001021915.1<span class=w> </span>REPR<span class=w>    </span>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/021/915/GCA_001021915.1_ASM102191v1<span class=w>  </span>-<span class=w>   </span>YM15
Xanthomonas<span class=w> </span>oryzae<span class=w> </span>pv.<span class=w> </span>oryzicola<span class=w>    </span><span class=m>129394</span><span class=w>  </span>PRJNA237971<span class=w> </span><span class=m>237971</span><span class=w>  </span>Proteobacteria<span class=w>  </span>Gammaproteobacteria<span class=w> </span><span class=m>5</span>.0801<span class=w>  </span><span class=m>64</span><span class=w>  </span>chromosome:NZ_CP007221.1/CP007221.1<span class=w> </span>-<span class=w>   </span><span class=m>1</span><span class=w>   </span><span class=m>4690</span><span class=w>    </span><span class=m>3924</span><span class=w>    </span><span class=m>2015</span>/03/04<span class=w>  </span><span class=m>2022</span>/01/21<span class=w>  </span>Complete<span class=w> </span>Genome<span class=w> </span>Cornell<span class=w> </span>University<span class=w>  </span>SAMN02640212<span class=w>    </span>GCA_000940825.1<span class=w> </span>-<span class=w>   </span>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/940/825/GCA_000940825.1_ASM94082v1<span class=w>   </span><span class=m>27148456</span><span class=w>    </span>CFBP7342
Xanthomonas<span class=w> </span>oryzae<span class=w> </span>pv.<span class=w> </span>oryzicola<span class=w>    </span><span class=m>129394</span><span class=w>  </span>PRJNA280380<span class=w> </span><span class=m>280380</span><span class=w>  </span>Proteobacteria<span class=w>  </span>Gammaproteobacteria<span class=w> </span><span class=m>5</span>.01777<span class=w> </span><span class=m>63</span>.9<span class=w>    </span>chromosome:NZ_CP011959.1/CP011959.1<span class=w> </span>-<span class=w>   </span><span class=m>1</span><span class=w>   </span><span class=m>4564</span><span class=w>    </span><span class=m>3815</span><span class=w>    </span><span class=m>2015</span>/06/30<span class=w>  </span><span class=m>2021</span>/11/23<span class=w>  </span>Complete<span class=w> </span>Genome<span class=w> </span>Bogdanove<span class=w> </span>Lab,<span class=w> </span>Plant<span class=w> </span>Pathology<span class=w> </span>and<span class=w> </span>Plant-Microbe<span class=w> </span>Biology,<span class=w> </span>Cornell<span class=w> </span>University<span class=w>    </span>SAMN03612287<span class=w>    </span>GCA_001042835.1<span class=w> </span>-<span class=w>   </span>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/042/835/GCA_001042835.1_ASM104283v1<span class=w>  </span>-<span class=w>   </span>CFBP7341

wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/021/915/GCA_001021915.1_ASM102191v1/GCA_001021915.1_ASM102191v1_genomic.fna.gz<span class=w> </span>-O<span class=w> </span>Xoc_GCA_001021915.fna.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/940/825/GCA_000940825.1_ASM94082v1/GCA_000940825.1_ASM94082v1_genomic.fna.gz<span class=w> </span>-O<span class=w> </span>Xoc_GCA_000940825.fna.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/042/835/GCA_001042835.1_ASM104283v1/GCA_001042835.1_ASM104283v1_genomic.fna.gz<span class=w> </span>-O<span class=w> </span>Xoc_GCA_001042835.fna.gz

wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/021/915/GCA_001021915.1_ASM102191v1/GCA_001021915.1_ASM102191v1_genomic.gbff.gz<span class=w> </span>-O<span class=w> </span>Xoc_GCA_001021915.gbff.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/940/825/GCA_000940825.1_ASM94082v1/GCA_000940825.1_ASM94082v1_genomic.gbff.gz<span class=w> </span>-O<span class=w> </span>Xoc_GCA_000940825.gbff.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/042/835/GCA_001042835.1_ASM104283v1/GCA_001042835.1_ASM104283v1_genomic.gbff.gz<span class=w> </span>-O<span class=w> </span>Xoc_GCA_001042835.gbff.gz
</code></pre></div> </details> <p>Do the same for retrieving the first 3 complete genomes of Xoc (Xanthomonas oryzae pv. oryzicola) with prefix Xoc_</p> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>grep<span class=w> </span><span class=s2>&quot;Xanthomonas oryzae pv. oryzae&quot;</span><span class=w> </span>prokaryotes.txt<span class=w> </span><span class=p>|</span><span class=w> </span>grep<span class=w> </span><span class=s1>&#39;chromosome:&#39;</span><span class=w> </span><span class=p>|</span><span class=w> </span>head<span class=w> </span>-3

Xanthomonas<span class=w> </span>oryzae<span class=w> </span>pv.<span class=w> </span>oryzae<span class=w>   </span><span class=m>64187</span><span class=w>   </span>PRJNA485465<span class=w> </span><span class=m>485465</span><span class=w>  </span>Proteobacteria<span class=w>  </span>Gammaproteobacteria<span class=w> </span><span class=m>4</span>.99067<span class=w> </span><span class=m>63</span>.7<span class=w>    </span>chromosome:NZ_CP031697.1/CP031697.1<span class=w> </span>-<span class=w>   </span><span class=m>1</span><span class=w>   </span><span class=m>4647</span><span class=w>    </span><span class=m>3652</span><span class=w>    </span><span class=m>2019</span>/02/04<span class=w>  </span><span class=m>2022</span>/03/27<span class=w>  </span>Complete<span class=w> </span>Genome<span class=w> </span>Martin<span class=w> </span>Luther<span class=w> </span>University<span class=w> </span>Halle-Wittenberg<span class=w>   </span>SAMN09791860<span class=w>    </span>GCA_004136375.1<span class=w> </span>-<span class=w>   </span>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/004/136/375/GCA_004136375.1_ASM413637v1<span class=w>  </span>-<span class=w>   </span>ICMP3125
Xanthomonas<span class=w> </span>oryzae<span class=w> </span>pv.<span class=w> </span>oryzae<span class=w>   </span><span class=m>64187</span><span class=w>   </span>PRJNA269487<span class=w> </span><span class=m>269487</span><span class=w>  </span>Proteobacteria<span class=w>  </span>Gammaproteobacteria<span class=w> </span><span class=m>5</span>.09305<span class=w> </span><span class=m>63</span>.7<span class=w>    </span>chromosome:NZ_CP040687.1/CP040687.1<span class=w> </span>-<span class=w>   </span><span class=m>1</span><span class=w>   </span><span class=m>4757</span><span class=w>    </span><span class=m>3787</span><span class=w>    </span><span class=m>2019</span>/05/28<span class=w>  </span><span class=m>2022</span>/03/20<span class=w>  </span>Complete<span class=w> </span>Genome<span class=w> </span>Bacterial<span class=w> </span>Genomics<span class=w> </span>and<span class=w> </span>Evolution<span class=w> </span>Laboratory,<span class=w> </span>CSIR-Institute<span class=w> </span>of<span class=w> </span>Microbial<span class=w> </span>Technology,<span class=w> </span>Sector<span class=w> </span><span class=m>39</span>-A,<span class=w> </span>Chandigarh,<span class=w> </span>India<span class=w> </span>SAMN03252586<span class=w>    </span>GCA_001929235.2<span class=w> </span>-<span class=w>   </span>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/929/235/GCA_001929235.2_ASM192923v2<span class=w>  </span>-<span class=w>   </span>IXO1088
Xanthomonas<span class=w> </span>oryzae<span class=w> </span>pv.<span class=w> </span>oryzae<span class=w>   </span><span class=m>64187</span><span class=w>   </span>PRJNA497605<span class=w> </span><span class=m>497605</span><span class=w>  </span>Proteobacteria<span class=w>  </span>Gammaproteobacteria<span class=w> </span><span class=m>5</span>.05039<span class=w> </span><span class=m>63</span>.7<span class=w>    </span>chromosome:NZ_CP033192.1/CP033192.1<span class=w> </span>-<span class=w>   </span><span class=m>1</span><span class=w>   </span><span class=m>4713</span><span class=w>    </span><span class=m>3695</span><span class=w>    </span><span class=m>2019</span>/03/19<span class=w>  </span><span class=m>2022</span>/04/07<span class=w>  </span>Complete<span class=w> </span>Genome<span class=w> </span>University<span class=w> </span>of<span class=w> </span>Florida<span class=w>   </span>SAMN10261817<span class=w>    </span>GCA_004355825.1<span class=w> </span>-<span class=w>   </span>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/004/355/825/GCA_004355825.1_ASM435582v1<span class=w>  </span>-<span class=w>   </span>NX0260

wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/004/136/375/GCA_004136375.1_ASM413637v1/GCA_004136375.1_ASM413637v1_genomic.fna.gz<span class=w> </span>-O<span class=w> </span>Xoo_GCA_004136375.fna.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/929/235/GCA_001929235.2_ASM192923v2/GCA_001929235.2_ASM192923v2_genomic.fna.gz<span class=w> </span>-O<span class=w> </span>Xoo_GCA_001929235.fna.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/004/355/825/GCA_004355825.1_ASM435582v1/GCA_004355825.1_ASM435582v1_genomic.fna.gz<span class=w> </span>-O<span class=w> </span>Xoo_GCA_004355825.fna.gz

wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/004/136/375/GCA_004136375.1_ASM413637v1/GCA_004136375.1_ASM413637v1_genomic.gbff.gz<span class=w> </span>-O<span class=w> </span>Xoo_GCA_004136375.gbff.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/929/235/GCA_001929235.2_ASM192923v2/GCA_001929235.2_ASM192923v2_genomic.gbff.gz<span class=w> </span>-O<span class=w> </span>Xoo_GCA_001929235.gbff.gz
wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/004/355/825/GCA_004355825.1_ASM435582v1/GCA_004355825.1_ASM435582v1_genomic.gbff.gz<span class=w> </span>-O<span class=w> </span>Xoo_GCA_004355825.gbff.gz
</code></pre></div> </details> <h2 id=convert-genbank-annotation-files-to-gff-format><strong>Convert GenBank annotation files to GFF format</strong><a class=headerlink href=#convert-genbank-annotation-files-to-gff-format title="Permanent link">#</a></h2> <p>We will make use of Perl scripts for converting genbank files to GFF format: <a href=https://github.com/bioperl/bioperl-live.git>https://github.com/bioperl/bioperl-live.git</a></p> <p>Clone this repository and run the dedicated bp_genbank2gff3 tool on each of the 6 Xanthomonas genomes.</p> <div class=highlight><pre><span></span><code>git clone https://github.com/bioperl/bioperl-live.git
chmod 755 bioperl-live/bin/*
</code></pre></div> <div class=highlight><pre><span></span><code>perl bioperl-live/bin/bp_genbank2gff3 Xoo_GCA_004136375.gbff.gz
perl bioperl-live/bin/bp_genbank2gff3 Xoo_GCA_001929235.gbff.gz
perl bioperl-live/bin/bp_genbank2gff3 Xoo_GCA_004355825.gbff.gz
perl bioperl-live/bin/bp_genbank2gff3 Xoc_GCA_001021915.gbff.gz
perl bioperl-live/bin/bp_genbank2gff3 Xoc_GCA_000940825.gbff.gz
perl bioperl-live/bin/bp_genbank2gff3 Xoc_GCA_001042835.gbff.gz
</code></pre></div> <div class=highlight><pre><span></span><code>mv<span class=w> </span>Xo*gff<span class=w> </span>annotations
</code></pre></div> <p>Have a look into the content of a GFF file.</p> <h2 id=comparison-between-2-genomes-with-minimap2-and-circos><strong>Comparison between 2 genomes (with Minimap2 and circos)</strong><a class=headerlink href=#comparison-between-2-genomes-with-minimap2-and-circos title="Permanent link">#</a></h2> <p>Compare 2 Xoo</p> <div class=highlight><pre><span></span><code>gunzip Xoo*fna.gz
</code></pre></div> <div class=highlight><pre><span></span><code>ls Xoo*fna
</code></pre></div> <p>Using minimap2, compare 2 genomes of Xoo (Xanthomonas oryzae pv.oryzae): Xoo_GCA_001929235 and Xoo_GCA_004136375. </p> <p>Generate an output in paf format named "minimap2.paf".</p> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>bioinfo/minimap2/2.24
minimap2<span class=w> </span>Xoo_GCA_001929235.fna<span class=w> </span>Xoo_GCA_004136375.fna<span class=w> </span>-o<span class=w> </span>minimap2.paf
</code></pre></div> </details> <div class="admonition question"> <p class=admonition-title>Exercises</p> <pre><code>Using awk command applied on the paf output of minimap2, we can try to filter matches between the two genomes in oreder to visualize connections with circos. Generate a new file named "links.txt" that contains links between the two genomes. 
On the other hand, in order to highlight inversion, do the same for matches that are inverted between the two genomes.
</code></pre> </div> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>awk<span class=w> </span><span class=o>{</span><span class=s1>&#39;if (($4-$3)&gt;100000 &amp;&amp; ($5 ==&quot;+&quot;))print NR-1&quot; &quot;$1&quot; &quot;$3&quot; &quot;$4&quot;\n&quot;NR-1&quot; &quot;$6&quot; &quot;$8&quot; &quot;$9&#39;</span><span class=o>}</span><span class=w> </span>minimap2.paf<span class=w> </span>&gt;links.txt
awk<span class=w> </span><span class=o>{</span><span class=s1>&#39;if (($4-$3)&gt;100000 &amp;&amp; ($5 ==&quot;-&quot;))print NR-1&quot; &quot;$1&quot; &quot;$3&quot; &quot;$4&quot;\n&quot;NR-1&quot; &quot;$6&quot; &quot;$9&quot; &quot;$8&#39;</span><span class=o>}</span><span class=w> </span>minimap2.paf<span class=w> </span>&gt;links_inverted.txt
</code></pre></div> </details> <p>Visualize matches between genomes with Circos</p> <p>We will use and adapt a template for circos configuration file for visualizing circos with links.</p> <div class=highlight><pre><span></span><code>cp -rf /scratch3/TP_bcg/$USER/training_bcg/training_bacterial_comparative_genomics/data/circos2.conf circos2.conf
</code></pre></div> <p>Edit the Circos configuration file to add informations about links</p> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>&lt;link<span class=w> </span>segdup-bundle34&gt;
<span class=nv>z</span><span class=w>            </span><span class=o>=</span><span class=w> </span><span class=m>50</span>
<span class=nv>color</span><span class=w>        </span><span class=o>=</span><span class=w> </span>nyt_red
<span class=nv>thickness</span><span class=w>    </span><span class=o>=</span><span class=w> </span><span class=m>3</span>
<span class=nv>file</span><span class=w>         </span><span class=o>=</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/public_genomes/links.txt
<span class=nv>bezier_radius_purity</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span>.2
<span class=nv>ribbon</span><span class=w>      </span><span class=o>=</span><span class=w> </span>yes
<span class=nv>crest</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span>
&lt;/link&gt;

&lt;link<span class=w> </span>segdup-bundle35&gt;
<span class=nv>z</span><span class=w>            </span><span class=o>=</span><span class=w> </span><span class=m>50</span>
<span class=nv>color</span><span class=w>        </span><span class=o>=</span><span class=w> </span>nyt_blue
<span class=nv>thickness</span><span class=w>    </span><span class=o>=</span><span class=w> </span><span class=m>3</span>
<span class=nv>file</span><span class=w>         </span><span class=o>=</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/public_genomes/links_inverted.txt
<span class=nv>bezier_radius_purity</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span>.2
<span class=nv>ribbon</span><span class=w>      </span><span class=o>=</span><span class=w> </span>yes
<span class=nv>crest</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span>
&lt;/link&gt;
</code></pre></div> </details> <p>Edit a karyotype file with the two genomes and their size</p> <p>Make one of the genome in the opposite way by entering this line in the Circos configuration file</p> <p>chromosomes_reverse = CP040687.1</p> <p>Launch circos for visualization</p> <div class=highlight><pre><span></span><code>circos/bin/circos<span class=w> </span>-conf<span class=w> </span>circos2.conf
</code></pre></div> <p><img alt src=../images/circos2.png></p> <p>Alternatively, you can align the two Xoo genomes with the web application DGenies for displaying dot plot large genomes in an interactive, efficient and simple way <a href=https://dgenies.toulouse.inra.fr/run>https://dgenies.toulouse.inra.fr/run</a></p> <p>Do the same process for the comparing one Xoo and one Xoc</p> <h2 id=compare-genome-similarity-using-average-nucleotide-identity-ani-fastani><strong>Compare genome similarity using Average Nucleotide Identity (ANI) (fastANI)</strong><a class=headerlink href=#compare-genome-similarity-using-average-nucleotide-identity-ani-fastani title="Permanent link">#</a></h2> <p>Go to the directory "public_genomes"</p> <div class=highlight><pre><span></span><code><span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/public_genomes
</code></pre></div> <p>Create a file that contains the list of genomes to be taken as input</p> <div class=highlight><pre><span></span><code>ls Xo*fna* &gt;list_genomes.txt
</code></pre></div> <div class=highlight><pre><span></span><code>module load bioinfo/fastani/1.33
fastANI --rl list_genomes.txt --ql list_genomes.txt -o fastani.out --matrix
</code></pre></div> <p>Check that genomes are closer within the same pathovar than between pathovars.</p> <div class="admonition question"> <p class=admonition-title>Questions</p> <pre><code>Which are the closest genomes?
</code></pre> </div> <details class=example> <summary>Solution</summary> <p><div class=highlight><pre><span></span><code>more<span class=w> </span>fastani.out.matrix<span class=w> </span>
</code></pre></div> <div class=highlight><pre><span></span><code>6  
Xoc_GCA_000940825.fna  
Xoc_GCA_001021915.fna   99.149597  
Xoc_GCA_001042835.fna   99.411034   99.196701  
Xoo_GCA_001929235.fna   97.226959   97.219826   97.288483  
Xoo_GCA_004136375.fna   97.313660   97.230530   97.365417   99.582520  
Xoo_GCA_004355825.fna   97.236938   97.235214   97.313324   99.704956   99.590271  
</code></pre></div></p> </details> <p>How to visualize ANI as heatmap matrix?</p> <div class=highlight><pre><span></span><code>cp<span class=w> </span>-rf<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/training_bacterial_comparative_genomics/scripts/convertANI.pl<span class=w> </span>.
cp<span class=w> </span>-rf<span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg/training_bacterial_comparative_genomics/scripts/heatmaply.R<span class=w> </span>.
perl<span class=w> </span>convertANI.pl<span class=w> </span>fastani.out.matrix<span class=w> </span>&gt;fastani.out.matrix.completed
</code></pre></div> <p>Install using the R interface the library optparse and heatmaply</p> <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>bioinfo/R/4.2.2
R
</code></pre></div> <div class=highlight><pre><span></span><code><span class=nf>install.packages</span><span class=p>(</span><span class=s>&quot;optparse&quot;</span><span class=p>)</span>
<span class=nf>install.packages</span><span class=p>(</span><span class=s>&quot;heatmaply&quot;</span><span class=p>)</span>
<span class=nf>quit</span><span class=p>()</span>
</code></pre></div> <p>Launch the R script to generate a figure showing a heatmap of ANI values</p> <div class=highlight><pre><span></span><code>Rscript<span class=w> </span>heatmaply.R<span class=w> </span>-f<span class=w> </span>fastani.out.matrix.completed
</code></pre></div> <p>It generates a HTML output showing a heatmap of ANI values for each genome pairwise comparison.</p> <p><img alt src=../images/ani.PNG></p> <p>Try to make the same process including our newly assembled genome</p> <h2 id=gene-content-comparison-pangenome-analysis-using-roary><strong>Gene content comparison / Pangenome analysis (using Roary)</strong><a class=headerlink href=#gene-content-comparison-pangenome-analysis-using-roary title="Permanent link">#</a></h2> <p>Have a look to the help page of the roary program.</p> <div class=highlight><pre><span></span><code><span class=nb>cd</span><span class=w> </span>/scratch3/TP_bcg/<span class=nv>$USER</span>/training_bcg
module<span class=w> </span>load<span class=w> </span>bioinfo/roary/3.12.0
roary<span class=w> </span>--help
</code></pre></div> <p>Launch roary for comparing the gene content of the 6 genomes.</p> <div class=highlight><pre><span></span><code>roary -f roary_out public_genomes/*gff
</code></pre></div> <p>Look at the content of the roary output:</p> <ul> <li> <p>Gene presence absence file</p> </li> <li> <p>summary statistics</p> </li> </ul> <div class=highlight><pre><span></span><code>head -10 roary_out/gene_presence_absence.csv  
&quot;Gene&quot;,&quot;Non-unique Gene name&quot;,&quot;Annotation&quot;,&quot;No. isolates&quot;,&quot;No. sequences&quot;,&quot;Avg sequences per isolate&quot;,&quot;Genome Fragment&quot;,&quot;Order within Fragment&quot;,&quot;Accessory Fragment&quot;,&quot;Accessory Order with Fragment&quot;,&quot;QC&quot;,&quot;Min group size nuc&quot;,&quot;Max group size nuc&quot;,&quot;Avg group size nuc&quot;,&quot;Xoc_GCA_000940825.gbff.gz&quot;,&quot;Xoc_GCA_001021915.gbff.gz&quot;,&quot;Xoc_GCA_001042835.gbff.gz&quot;,&quot;Xoo_GCA_001929235.gbff.gz&quot;,&quot;Xoo_GCA_004136375.gbff.gz&quot;,&quot;Xoo_GCA_004355825.gbff.gz&quot;
&quot;IXO1088_008755&quot;,&quot;&quot;,&quot;response regulator&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1683&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;404&quot;,&quot;404&quot;,&quot;404&quot;,&quot;BE73_06245.p01&quot;,&quot;FE36_14125.p01&quot;,&quot;ACU17_06495.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6321&quot;,&quot;4e95a45ca635958849417c3d368d746d_7522&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11970&quot;
&quot;IXO1088_008760&quot;,&quot;&quot;,&quot;hypothetical protein&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1682&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;497&quot;,&quot;497&quot;,&quot;497&quot;,&quot;BE73_06250.p01&quot;,&quot;FE36_14120.p01&quot;,&quot;ACU17_06500.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6325&quot;,&quot;4e95a45ca635958849417c3d368d746d_7526&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11966&quot;
&quot;EBA21_16420&quot;,&quot;&quot;,&quot;hypothetical protein&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1673&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;512&quot;,&quot;512&quot;,&quot;512&quot;,&quot;BE73_06290.p01&quot;,&quot;FE36_14080.p01&quot;,&quot;ACU17_06540.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6351&quot;,&quot;4e95a45ca635958849417c3d368d746d_7554&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11934&quot;
&quot;pqqD&quot;,&quot;&quot;,&quot;pyrroloquinoline quinone biosynthesis peptide chaperone PqqD&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1671&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;278&quot;,&quot;278&quot;,&quot;278&quot;,&quot;BE73_06300.p01&quot;,&quot;FE36_14070.p01&quot;,&quot;ACU17_06550.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6359&quot;,&quot;4e95a45ca635958849417c3d368d746d_7562&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11926&quot;
&quot;pqqB&quot;,&quot;&quot;,&quot;pyrroloquinoline quinone biosynthesis protein PqqB&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1669&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;899&quot;,&quot;899&quot;,&quot;899&quot;,&quot;BE73_06310.p01&quot;,&quot;FE36_14060.p01&quot;,&quot;ACU17_06560.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6367&quot;,&quot;4e95a45ca635958849417c3d368d746d_7570&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11918&quot;
&quot;DZA53_10525&quot;,&quot;&quot;,&quot;hypothetical protein&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1664&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;524&quot;,&quot;524&quot;,&quot;524&quot;,&quot;BE73_06335.p01&quot;,&quot;FE36_14035.p01&quot;,&quot;ACU17_06585.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6391&quot;,&quot;4e95a45ca635958849417c3d368d746d_7594&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11894&quot;
&quot;DZA53_10545&quot;,&quot;&quot;,&quot;energy transducer TonB&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1661&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;875&quot;,&quot;875&quot;,&quot;875&quot;,&quot;BE73_06355.p01&quot;,&quot;FE36_14015.p01&quot;,&quot;ACU17_06605.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6407&quot;,&quot;4e95a45ca635958849417c3d368d746d_7610&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11878&quot;
&quot;DZA53_10580&quot;,&quot;&quot;,&quot;chemotaxis protein&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1654&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;1202&quot;,&quot;1208&quot;,&quot;1205&quot;,&quot;BE73_06390.p01&quot;,&quot;FE36_13980.p01&quot;,&quot;ACU17_06640.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6435&quot;,&quot;4e95a45ca635958849417c3d368d746d_7638&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11850&quot;
&quot;ACU17_06675&quot;,&quot;&quot;,&quot;copper homeostasis protein CutC&quot;,&quot;6&quot;,&quot;6&quot;,&quot;1&quot;,&quot;1&quot;,&quot;1646&quot;,&quot;&quot;,&quot;&quot;,&quot;&quot;,&quot;731&quot;,&quot;731&quot;,&quot;731&quot;,&quot;BE73_06425.p01&quot;,&quot;FE36_13950.p01&quot;,&quot;ACU17_06675.p01&quot;,&quot;10bd397a0f88aca11976f4a4f6631125_6467&quot;,&quot;4e95a45ca635958849417c3d368d746d_7672&quot;,&quot;cc532a0030968c6887cea9dec5a101c0_11818&quot;
</code></pre></div> <div class=highlight><pre><span></span><code>more<span class=w> </span>roary_out/summary_statistics.txt<span class=w>  </span>
Core<span class=w> </span>genes<span class=w>  </span><span class=o>(</span><span class=m>99</span>%<span class=w> </span>&lt;<span class=o>=</span><span class=w> </span>strains<span class=w> </span>&lt;<span class=o>=</span><span class=w> </span><span class=m>100</span>%<span class=o>)</span><span class=w>    </span><span class=m>2565</span>
Soft<span class=w> </span>core<span class=w> </span>genes<span class=w> </span><span class=o>(</span><span class=m>95</span>%<span class=w> </span>&lt;<span class=o>=</span><span class=w> </span>strains<span class=w> </span>&lt;<span class=w> </span><span class=m>99</span>%<span class=o>)</span><span class=w>  </span><span class=m>0</span>
Shell<span class=w> </span>genes<span class=w> </span><span class=o>(</span><span class=m>15</span>%<span class=w> </span>&lt;<span class=o>=</span><span class=w> </span>strains<span class=w> </span>&lt;<span class=w> </span><span class=m>95</span>%<span class=o>)</span><span class=w>  </span><span class=m>3763</span>
Cloud<span class=w> </span>genes<span class=w> </span><span class=o>(</span><span class=m>0</span>%<span class=w> </span>&lt;<span class=o>=</span><span class=w> </span>strains<span class=w> </span>&lt;<span class=w> </span><span class=m>15</span>%<span class=o>)</span><span class=w>   </span><span class=m>0</span>
Total<span class=w> </span>genes<span class=w> </span><span class=o>(</span><span class=m>0</span>%<span class=w> </span>&lt;<span class=o>=</span><span class=w> </span>strains<span class=w> </span>&lt;<span class=o>=</span><span class=w> </span><span class=m>100</span>%<span class=o>)</span><span class=w> </span><span class=m>6328</span>
</code></pre></div> <p>In order to explore and represent graphs with roary outputs, we will make use of scripts available in Roary repository.</p> <div class=highlight><pre><span></span><code>git clone https://github.com/sanger-pathogens/Roary.git
</code></pre></div> <div class=highlight><pre><span></span><code>python<span class=w> </span>Roary/contrib/roary_plots/roary_plots.py<span class=w> </span>roary_out/accessory_binary_genes.fa.newick<span class=w> </span>roary_out/gene_presence_absence.csv
</code></pre></div> <div class=highlight><pre><span></span><code>query_pan_genome<span class=w> </span>-g<span class=w> </span>roary_out/clustered_proteins<span class=w> </span>-a<span class=w> </span>difference<span class=w> </span>--input_set_one<span class=w> </span>public_genomes/Xoc_GCA_000940825.gbff.gz.gff,public_genomes/Xoc_GCA_001021915.gbff.gz.gff,public_genomes/Xoc_GCA_001021915.gbff.gz.gff<span class=w> </span>--input_set_two<span class=w> </span>public_genomes/Xoo_GCA_001929235.gbff.gz.gff,public_genomes/Xoo_GCA_004136375.gbff.gz.gff,public_genomes/Xoo_GCA_004355825.gbff.gz.gff
</code></pre></div> <p><img alt src=../images/pangenome_pie.png> <img alt src=../images/pangenome_matrix.png></p> <h2 id=association-tests-with-metadata-pan-gwas-with-scoary><strong>Association tests with metadata (Pan-GWAS) (with Scoary)</strong><a class=headerlink href=#association-tests-with-metadata-pan-gwas-with-scoary title="Permanent link">#</a></h2> <p>Scoary is designed to take the gene_presence_absence.csv file from Roary as well as a traits file created by the user and calculate the assocations between all genes in the accessory genome and the traits. It reports a list of genes sorted by strength of association per trait. </p> <p>For this task, we will use a larger file precalculated with Roary for the comparison of 90 strains. For each strain, we collected the pathovar information and we will use it to define genes potentially associated with pathovar Xoo/Xoc.</p> <div class=highlight><pre><span></span><code>cp -rf /scratch3/TP_bcg/$USER/training_bcg/training_bacterial_comparative_genomics/data/pangenome_100xantho /scratch3/TP_bcg/$USER/training_bcg
cd /scratch3/TP_bcg/$USER/training_bcg/pangenome_100xantho
</code></pre></div> <p>Look into the input files gene_presence_absence.csv and traits.txt </p> <div class=highlight><pre><span></span><code>more gene_presence_absence.csv
</code></pre></div> <div class=highlight><pre><span></span><code>more traits.txt
</code></pre></div> <p>Look at the help page of scoary and run it for defining genes associated with pathovar.</p> <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>bioinfo/scoary/1.6.16
scoary<span class=w> </span>-t<span class=w> </span>traits.txt<span class=w> </span>-g<span class=w> </span>gene_presence_absence.csv
</code></pre></div> <div class="admonition question"> <p class=admonition-title>Questions</p> <pre><code>Considering a p-value threshold 0.01, how many gene clusters are associated with pathovar?  
How many clusters are specifically present in pathovar oryzae (absent in oryzicola)?  
How many clusters are specifically present in pathovar oryzicola (absent in oryzae)?
</code></pre> </div> <details class=example> <summary>Solution</summary> <div class=highlight><pre><span></span><code>awk -F &quot;\&quot;,\&quot;&quot; {&#39;if ($12&lt;0.01)print $_&#39;} pathovar_*.results.csv | wc -l
</code></pre></div> <div class=highlight><pre><span></span><code>awk<span class=w> </span>-F<span class=w> </span><span class=s2>&quot;\&quot;,\&quot;&quot;</span><span class=w> </span><span class=o>{</span><span class=s1>&#39;if ($4==0 &amp;&amp; $5==75)print $_&#39;</span><span class=o>}</span><span class=w> </span>pathovar_*.results.csv<span class=w> </span><span class=p>|</span><span class=w> </span>wc<span class=w> </span>-l
</code></pre></div> <div class=highlight><pre><span></span><code>awk<span class=w> </span>-F<span class=w> </span><span class=s2>&quot;\&quot;,\&quot;&quot;</span><span class=w> </span><span class=o>{</span><span class=s1>&#39;if ($4==15 &amp;&amp; $5==0)print $_&#39;</span><span class=o>}</span><span class=w> </span>pathovar_*.results.csv<span class=w> </span><span class=p>|</span><span class=w> </span>wc<span class=w> </span>-l
</code></pre></div> </details> <h2 id=explore-pan-genome-with-panexplorer-web-application><strong>Explore pan-genome with PanExplorer web application</strong><a class=headerlink href=#explore-pan-genome-with-panexplorer-web-application title="Permanent link">#</a></h2> <p>Upload your 6 genomes into the web application and choose the PanACoTA software for pangenome reconstruction. <a href=https://panexplorer.southgreen.fr/ >https://panexplorer.southgreen.fr/</a></p> <p><img alt src=../images/panex_capture.PNG></p> <div class="admonition question"> <p class=admonition-title>Exercises</p> <pre><code>How many genes are strain-specific of each of the 3 Xoo strains?
</code></pre> </div> <h2 id=construct-a-pangenome-graph-using-minigraph2><strong>Construct a pangenome graph (using minigraph2)</strong><a class=headerlink href=#construct-a-pangenome-graph-using-minigraph2 title="Permanent link">#</a></h2> <p>Pangenome analysis based on gene clustering approach will result in a pangene atlas, construction of a presence/absence matrix of genes. </p> <p>Pangenome can also be reconstructed directly on genomic sequences using pangenome graph approach using minigraph2.</p> <div class=highlight><pre><span></span><code>cd /scratch3/TP_bcg/$USER/training_bcg/public_genomes
module load bioinfo/minigraph/0.15
minigraph -xggs -t16 Xoo_GCA_001929235.fna Xoo_GCA_004136375.fna &gt;Xoo_pangenome.gfa
</code></pre></div> <p>Convert GFA into FASTA of the pangenome</p> <div class=highlight><pre><span></span><code>module load bioinfo/gfatools/0.5
gfatools gfa2fa -s Xoo_pangenome.gfa &gt;Xoo_pangenome.fa
</code></pre></div> <p>Download the bandage software and load the GFA pangenome graph to be visualized</p> <p><img alt src=../images/bandage_capture.PNG></p> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../.. class="md-footer__link md-footer__link--prev" aria-label="Previous: Home"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Home </div> </div> </a> <a href=../lecture/ class="md-footer__link md-footer__link--next" aria-label="Next: Lecture"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Lecture </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "header.autohide", "navigation.footer", "navigation.indexes", "navigation.tabs", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../../assets/javascripts/bundle.13a4f30d.min.js></script> </body> </html>