# Bacterial comparative genomics

![](pages/images/achievement-agreement-arms-1068523.jpg)

## Course overview

<img src="https://www.svgrepo.com/show/305241/github.svg"
    width="20" height="20"/>
    [GitHub repository]( {{config.repo_url}})
<img src="https://www.svgrepo.com/show/20800/event-date-and-time-symbol.svg"
    width="20" height="20"/>
    8 - 9 June, 2023

This course aims at discovering the main bioinformatics for comparing bacterial genomes, from pairwise comparison to a large scale pan-genome analysis.


## Course content and learning outcomes

The following topics and tools are covered in the course:


## Application

By accepting to participate in the course, you agree to follow the [Code of Conduct](pages/course-information/code-of-conduct.md).

## Schedule

## Schedule

8 June

* Genome assembly
* Identification of plasmid scaffolds
* Annotation (structural and functional annotation) of bacterial genomes
* Circos representation of genomes
* Pairwise comparison of genomes (ANI)


9 June:  
* Comparative analysis of gene content  
* PAV (Presence/Absence Variation))  
* Pan-genome/core-genome aspects  
* pan-GWAS  
* Pan-genome graph approach  


## Location

This course round is given on site. It will take place in the Badiane room at Agropolis.

## Course material

here lists all the information you need before the course starts. 

## Teachers

* Alexis Dereeper (course leader)
* Jacques Dainat
* Florian Charriat
* Antoni Exbrayat


## Contact

To contact us, please send a mail using the contact form available [here]({{contact}}).
